### PR TITLE
Adding Zone validation function for PowerFlex

### PR DESCRIPTION
**Pull request by anathoodell**
# Description
CSM-Operator Zone Validation function is added to validate zone configuration in a cluster of nodes and powerflex secret, prior to installation (pre-checks). If the Zone validation fails, the CSM object remains in a failed state.

# GitHub Issues
List the GitHub issues impacted by this PR:

| GitHub Issue # |
| -------------- |
| https://github.com/dell/csm/issues/1613 |

# Checklist:

- [x] I have performed a self-review of my own code to ensure there are no formatting, vetting, linting, or security issues
- [x] I have verified that new and existing unit tests pass locally with my changes
- [x] I have not allowed coverage numbers to degenerate
- [ ] I have maintained at least 90% code coverage
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have maintained backward compatibility
- [ ] I have executed the relevant end-to-end test scenarios

# How Has This Been Tested?
Unit tests added and run to maintain coverage numbers. Basic pflex E2E scenarios run as well.

- [x] Directly on a cluster with 2 nodes.
- [ ] Test B

if zone validation fails - this event is logged in the csm cr.
image

### CODE REVIEW

File: controllers/csm_controller.go [hunk 1]
```diff
@@ -1521,9 +1529,75 @@ func (r *ContainerStorageModuleReconciler) GetK8sClient() kubernetes.Interface {
...
+		for _, node := range matchingNodeList.Items {
+			log.Infof(" matching node found with node name %s", node.Name)
+		}
+		if len(matchingNodeList.Items) != numberOfNodes {
+			validZoneConfiguration = false
+			log.Errorw("Zone configuration mismatch ", len(matchingNodeList.Items))
+		}
+	}
+
+	return validZoneConfiguration, err
```
Comment (alexemc) [inline controllers/csm_controller.go:1602]: `validZoneConfiguration` is really not needed, `err == nil` would indicate a success.

File: controllers/csm_controller.go [hunk 2]
```diff
@@ -1521,9 +1529,75 @@ func (r *ContainerStorageModuleReconciler) GetK8sClient() kubernetes.Interface {
...
+	// a zone label. this constitutes a zone configuration failure.
+	// Determine how many arrays are configured and match that with the len(map)
+
+	// The labelKey, GetMatchingNodes with labelKey, value="", verify that the list
+	// returned is the same size as the current number of nodes on the cluster
+	// i.e. all nodes must have this zoning label key. if the number of nodes
+	// returned does not match the number of nodes on the cluster, then fail,
+	// otherwise return true
+	if firstKeyLabel != "" {
+		matchingNodeList, err := r.GetMatchingNodes(ctx, firstKeyLabel, "")
```
Comment (alexemc) [inline controllers/csm_controller.go:1587]: The function name is not too descriptive, it could be something like `GetNodesByLabel`.

**Review by alexemc — CHANGES_REQUESTED** [review]

File: controllers/csm_controller.go [hunk 3]
```diff
@@ -1334,6 +1334,11 @@ func (r *ContainerStorageModuleReconciler) PreChecks(ctx context.Context, cr *cs
 		if err != nil {
 			return fmt.Errorf("failed powerflex validation: %v", err)
 		}
+		// zoning initially applies only to pflex
+		err = r.ZoneValidation(ctx, cr, namespace)
```
Comment (JacobGros) [inline controllers/csm_controller.go:1338]: Do we need to pass in namespace here and in pre-checks?
If we have the cr, we should be able to get the namespace from 
cr.Namespace.
Like here: https://github.com/dell/csm-operator/blob/f54c83af80cc5d3428019ff6b78e36c4260dab17/controllers/csm_controller.go#L1371

File: controllers/csm_controller.go [hunk 4]
```diff
@@ -1521,9 +1529,75 @@ func (r *ContainerStorageModuleReconciler) GetK8sClient() kubernetes.Interface {
...
 	err := r.List(ctx, nodeList, opts...)
 
 	return nodeList, err
 }
+
+// Invalid zoning configurations:
+// 1. Some arrays have a zone label A, but one of the arrays has zone label B, or no label
+// 2. Array secret has zone label A, but one of the nodes has zone label B, or no label
+// 3. Some of the nodes have a zone label, but arrays have no zone labels.
+// 4. All of the arrays have a zone label, but nodes have no zone label.
```
[the remark this answers is not in the document]
Comment (JacobGros) [reply controllers/csm_controller.go:1546]: I think a node not having a zone key/label is a valid case; 
customer may not want the node to connect to any pflex array and in this case, 
the driver would pass MDM="" to the sdc init and the init container would start with no MDMs connected to it

File: controllers/csm_controller.go [hunk 5]
```diff
@@ -1524,12 +1529,68 @@ func (r *ContainerStorageModuleReconciler) GetK8sClient() kubernetes.Interface {
...
+
+	// Make sure all zone keys are the same, if not fail
+	var firstKeyLabel string
+	for key := range zones {
+		if firstKeyLabel == "" {
+			firstKeyLabel = key
+			continue
+		}
+		if key != firstKeyLabel {
+			return fmt.Errorf("detected mismatched Keys in zones")
```
Comment (alexemc) [inline controllers/csm_controller.go:1570]: As a user, I would like to know more specifically where is the misconfiguration: e.g. which array gave the error, or maybe a hint similar to "check the array info secret, zone key should be the same for all arrays".

File: pkg/drivers/powerflex.go [hunk 6]
```diff
@@ -375,7 +381,7 @@ func ExtractZonesFromSecret(ctx context.Context, kube client.Client, namespace s
 		}
 		err = yaml.Unmarshal(configs, &yamlConfig)
 		if err != nil {
-			return nil, fmt.Errorf("unable to unmarshal multi-array configuration[%v]", err)
+			return nil, fmt.Errorf("unable to unmarshal no Zone or Name or LabelKey found multi-array configuration[%v]", err)
```
Comment (JacobGros) [inline pkg/drivers/powerflex.go:384]: do we know if there's no zone or name or label key at this part?

File: controllers/csm_controller.go [hunk 7]
```diff
@@ -1524,12 +1529,68 @@ func (r *ContainerStorageModuleReconciler) GetK8sClient() kubernetes.Interface {
...
+	}
+	numberOfNodes := len(nodeList.Items)
+
+	zones, err := drivers.ExtractZones(ctx, cr, r.Client, namespace)
+	if err != nil {
+		return fmt.Errorf("ZoneValidation failed with error: %v", err)
+	}
+
+	// Make sure all zone keys are the same, if not fail
+	var firstKeyLabel string
```
Comment (JacobGros) [inline controllers/csm_controller.go:1563]: I don't think this check is working 
we build out zones map like this:
```
zonesMapData[configParam.SystemID] = configParam.Zone.Name
```
so here we're actually checking systemIDs here

Comment (JacobGros) [reply controllers/csm_controller.go:1563]: a check for label keys could be done in 
```ExtractZonesFromSecret``` as we parse through the secret

File: pkg/drivers/powerflex.go (hunk 6 above)
Comment (alexemc) [inline pkg/drivers/powerflex.go:384]: Is this right? I don't see labelKey defined in StorageArrayConfig and also zoning configuration is optional (at least for the backward compatibility) - we just need to check that all arrays have consistent configuration.

This function has a few errors formatted as `[%v]`, can you change it to `: %v` for consistency with our conventional formatting.

File: controllers/csm_controller.go [hunk 8]
```diff
@@ -1524,12 +1529,68 @@ func (r *ContainerStorageModuleReconciler) GetK8sClient() kubernetes.Interface {
...
+	// TBD: also check if there any arrays specified in the secret that do not have
+	// a zone label. this constitutes a zone configuration failure.
+	// Determine how many arrays are configured and match that with the len(map)
+
+	// The labelKey, GetMatchingNodes with labelKey, verify that the list
+	// returned is the same size as the current number of nodes on the cluster
+	// i.e. all nodes must have this zoning label key. if the number of nodes
+	// returned does not match the number of nodes on the cluster, then fail,
+	// otherwise return true
+	if firstKeyLabel != "" {
```
Comment (JacobGros) [inline controllers/csm_controller.go:1582]: we should probably double check w/ @tdawe but I think its ok for nodes not to have a label -> this means customer doesn't want driver to connect to any array on this node

**Review by JacobGros — CHANGES_REQUESTED** [review]

File: controllers/csm_controller.go (hunk 7 above)
Comment (alexemc) [reply controllers/csm_controller.go:1563]: `ExtractZoncesFromSecret` should only extract, validations should be done in `ZoneValidation`.

`ExtractZonesFromSecret` should return a slice of structs { systemId, zoneKey (optional), zoneValue (optional)} to give  `ZoneValidation` full information for validation.

File: controllers/csm_controller.go (hunk 8 above)
Comment (anathoodell) [reply controllers/csm_controller.go:1582]: that's different from what we understood.

Either all nodes in the cluster (except master) have zone labels, or none of them do. there's no in-between, at least that's my understanding.

Comment (tdawe) [reply controllers/csm_controller.go:1582]: The customer may only want the DaemonSet to be running on certain nodes (see option for this with the nodeSelector: https://github.com/dell/csm-operator/blob/main/samples/storage_csm_powerflex_v2130.yaml#L167). Therefore, the zoning label is not required to be on all nodes in the k8s cluster.

File: tests/shared/crclient/client.go [hunk 9]
```diff
@@ -124,12 +142,22 @@ func (f Client) listPodList(list *corev1.PodList) error {
...
 	for k, v := range f.Objects {
 		if k.Kind == "Node" {
-			node, ok := v.(*corev1.Node)
-			if ok {
-				list.Items = append(list.Items, *node)
+			node := *v.(*corev1.Node)
+			if label != "" {
+				for key := range node.ObjectMeta.Labels {
+					if label == key {
+						log.Infof("\tadding node name:%v to list matching label key \n", node.Name)
```
Comment (alexemc) [inline tests/shared/crclient/client.go:153]: `\n` is redundant. Also, matching which key? Do we need this debug log here at all?

File: tests/shared/crclient/client.go [hunk 10]
```diff
@@ -124,12 +142,22 @@ func (f Client) listPodList(list *corev1.PodList) error {
 	return nil
 }
 
-func (f Client) listNodeList(list *corev1.NodeList) error {
+func (f Client) listNodeList(list *corev1.NodeList, label string) error {
+	_, log := logger.GetNewContextWithLogger("0")
```
Comment (alexemc) [inline tests/shared/crclient/client.go:146]: You should probably just use `GetLogger`. Or don't log at all in this utility function.

**Comment by anathoodell** [comment]
Given the recent discussion around node validation with @tdawe, the scope of zone validation required is much reduced, only to the new pflex secret validation. 

therefore most of the new ZoneValidation code will disappear and the unit tests around it will also be quite simplified and reduced.

Given this rework - I will close this PR and raise a new one with the new simplified changes which will not do any node label validation as this as we agreed is too restrictive.

File: tests/shared/crclient/client.go (hunk 10 above)
Comment (anathoodell) [reply tests/shared/crclient/client.go:146]: I'll remove the logging here.

File: tests/shared/crclient/client.go (hunk 9 above)
Comment (anathoodell) [reply tests/shared/crclient/client.go:153]: removed.

File: tests/shared/crclient/client.go [hunk 11]
```diff
@@ -124,12 +142,22 @@ func (f Client) listPodList(list *corev1.PodList) error {
 	return nil
 }
 
-func (f Client) listNodeList(list *corev1.NodeList) error {
+func (f Client) listNodeList(list *corev1.NodeList, label string) error {
```
[the remark this answers is not in the document]
Comment (anathoodell) [reply tests/shared/crclient/client.go:145]: there's also a listPodList. keeping the same convention.

File: tests/shared/crclient/client.go [hunk 12]
```diff
@@ -124,12 +142,22 @@ func (f Client) listPodList(list *corev1.PodList) error {
...
+	_, log := logger.GetNewContextWithLogger("0")
 	for k, v := range f.Objects {
 		if k.Kind == "Node" {
-			node, ok := v.(*corev1.Node)
-			if ok {
-				list.Items = append(list.Items, *node)
+			node := *v.(*corev1.Node)
```
[the remark this answers is not in the document]
Comment (anathoodell) [reply tests/shared/crclient/client.go:149]: I put it back. thanks for that.

File: tests/shared/crclient/client.go [hunk 13]
```diff
@@ -107,7 +108,24 @@ func (f Client) List(ctx context.Context, list client.ObjectList, _ ...client.Li
 	case *corev1.PodList:
 		return f.listPodList(l)
 	case *corev1.NodeList:
-		return f.listNodeList(l)
+
+		var labelKey string
+		// Initialize ListOptions
+		listOpts := &client.ListOptions{}
+		// Apply each ListOption to listOpts
+		if opts != nil {
```
[the remark this answers is not in the document]
Comment (anathoodell) [reply tests/shared/crclient/client.go:116]: removed.

File: controllers/csm_controller_test.go [hunk 14]
```diff
@@ -2436,3 +2435,133 @@ func (suite *CSMControllerTestSuite) TestGetNodeLabels() {
...
+	sec := shared.MakeSecretPowerFlexMultiZone(csmName+"-config", suite.namespace, pFlexConfigVersion)
+	err = suite.fakeClient.Create(ctx, sec)
+	if err != nil {
+		panic(err)
+	}
+
+	err = reconciler.ZoneValidation(ctx, &csm, suite.namespace)
+	if err == nil {
+		panic(err)
+	}
```
[the remark this answers is not in the document]
Comment (anathoodell) [reply controllers/csm_controller_test.go:2566]: done.

File: controllers/csm_controller_test.go [hunk 15]
```diff
@@ -2436,3 +2435,133 @@ func (suite *CSMControllerTestSuite) TestGetNodeLabels() {
...
+	csm := shared.MakeCSM(csmName, suite.namespace, configVersion)
+	csm.Spec.Driver.CSIDriverType = csmv1.PowerFlex
+	csm.Spec.Driver.Common.Image = "image"
+	csm.Annotations[configVersionKey] = configVersion
+
+	csm.ObjectMeta.Finalizers = []string{CSMFinalizerName}
+	err := suite.fakeClient.Create(ctx, &csm)
+	if err != nil {
+		panic(err)
+	}
```
[the remark this answers is not in the document]
Comment (anathoodell) [reply controllers/csm_controller_test.go:2540]: done.

File: controllers/csm_controller_test.go [hunk 16]
```diff
@@ -2436,3 +2435,133 @@ func (suite *CSMControllerTestSuite) TestGetNodeLabels() {
...
```
[the remark this answers is not in the document]
Comment (anathoodell) [reply controllers/csm_controller_test.go:2449]: done.

File: controllers/csm_controller.go (hunk 8 above)
Comment (anathoodell) [reply controllers/csm_controller.go:1582]: PR closed and code removed as part of another PR.

File: pkg/drivers/powerflex.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (anathoodell) [reply pkg/drivers/powerflex.go:384]: this has been fixed in the next PR.

Comment (anathoodell) [reply pkg/drivers/powerflex.go:384]: removed.

File: controllers/csm_controller.go (hunk 3 above)
Comment (anathoodell) [reply controllers/csm_controller.go:1338]: makes sense. updated function.

File: pkg/drivers/powerflex.go (hunk 6 above)
Comment (anathoodell) [reply pkg/drivers/powerflex.go:384]: fixed. fixed the [%v] as well.

File: controllers/csm_controller.go (hunk 7 above)
Comment (anathoodell) [reply controllers/csm_controller.go:1563]: This code is slightly adjusted to find the labelKey and validate consistency across all arrays.

File: controllers/csm_controller_test.go [hunk 17]
```diff
@@ -2436,3 +2435,133 @@ func (suite *CSMControllerTestSuite) TestGetNodeLabels() {
...
+	// Add node4 that has a different zone label, and zone validate should fail
+	node4 := shared.MakeNode("node4", suite.namespace)
+	node4.Labels["my-special-zone-label"] = "mydrzone"
+	err = suite.fakeClient.Create(ctx, &node4)
+	assert.Nil(suite.T(), err)
+
+	err = reconciler.ZoneValidation(ctx, &csm, suite.namespace)
+	if err == nil {
+		panic(err)
+	}
```
[the remark this answers is not in the document]
Comment (anathoodell) [reply controllers/csm_controller_test.go:2527]: this code is removed given the new direction.

File: controllers/csm_controller.go (hunk 2 above)
Comment (anathoodell) [reply controllers/csm_controller.go]: just realized we no longer need this function.

File: controllers/csm_controller.go [hunk 18]
```diff
@@ -1521,9 +1529,75 @@ func (r *ContainerStorageModuleReconciler) GetK8sClient() kubernetes.Interface {
...
+
+	// The labelKey, GetMatchingNodes with labelKey, value="", verify that the list
+	// returned is the same size as the current number of nodes on the cluster
+	// i.e. all nodes must have this zoning label key. if the number of nodes
+	// returned does not match the number of nodes on the cluster, then fail,
+	// otherwise return true
+	if firstKeyLabel != "" {
+		matchingNodeList, err := r.GetMatchingNodes(ctx, firstKeyLabel, "")
+		if err != nil {
+			log.Errorw("Failed to retrieve list of nodes for label", firstKeyLabel)
```
[the remark this answers is not in the document]
Comment (anathoodell) [reply controllers/csm_controller.go]: code is removed.

File: controllers/csm_controller.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (anathoodell) [reply controllers/csm_controller.go]: removed.

File: controllers/csm_controller.go [hunk 19]
```diff
@@ -1328,6 +1328,14 @@ func (r *ContainerStorageModuleReconciler) PreChecks(ctx context.Context, cr *cs
 		if err != nil {
 			return fmt.Errorf("failed powerflex validation: %v", err)
 		}
+		// zoning initially applies only to pflex
+		zonesAreValid, err := r.ZoneValidation(ctx, cr, namespace)
+		if err != nil {
+			return fmt.Errorf("error during zone validation: %v", err)
+		}
+		if zonesAreValid == false {
+			return fmt.Errorf("failed zone validation.")
```
[the remark this answers is not in the document]
Comment (anathoodell) [reply controllers/csm_controller.go]: removed.

File: controllers/csm_controller.go [hunk 20]
```diff
@@ -1328,6 +1328,14 @@ func (r *ContainerStorageModuleReconciler) PreChecks(ctx context.Context, cr *cs
 		if err != nil {
 			return fmt.Errorf("failed powerflex validation: %v", err)
 		}
+		// zoning initially applies only to pflex
+		zonesAreValid, err := r.ZoneValidation(ctx, cr, namespace)
```
[the remark this answers is not in the document]
Comment (anathoodell) [reply controllers/csm_controller.go]: agreed. this is done.

File: controllers/csm_controller.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (anathoodell) [reply controllers/csm_controller.go:1546]: removed comments as rn they add no value.

File: controllers/csm_controller.go [hunk 21]
```diff
@@ -1521,9 +1529,75 @@ func (r *ContainerStorageModuleReconciler) GetK8sClient() kubernetes.Interface {
...
+	}
+
+	log.Infof("ZoneValidation zones returned from ExtractZones: %v", zones)
+	// Make sure all zone keys are the same, if not fail
+	i := 0
+	var firstKeyLabel string = ""
+	for key, value := range zones {
+		if i == 0 {
+			firstKeyLabel = key
+			log.Infof("firstKeyLabel %s\n", firstKeyLabel)
```
[the remark this answers is not in the document]
Comment (anathoodell) [reply controllers/csm_controller.go]: removed.

File: controllers/csm_controller.go [hunk 22]
```diff
@@ -1521,9 +1529,75 @@ func (r *ContainerStorageModuleReconciler) GetK8sClient() kubernetes.Interface {
...
+		log.Infof("ZoneValidation failed with error: %v", err)
+		return false, err
+	}
+
+	log.Infof("ZoneValidation zones returned from ExtractZones: %v", zones)
+	// Make sure all zone keys are the same, if not fail
+	i := 0
+	var firstKeyLabel string = ""
+	for key, value := range zones {
+		if i == 0 {
```
[the remark this answers is not in the document]
Comment (anathoodell) [reply controllers/csm_controller.go]: yes done.

File: controllers/csm_controller.go [hunk 23]
```diff
@@ -1521,9 +1529,75 @@ func (r *ContainerStorageModuleReconciler) GetK8sClient() kubernetes.Interface {
...
+			firstKeyLabel = key
+			log.Infof("firstKeyLabel %s\n", firstKeyLabel)
+			i = i + 1
+			continue
+		}
+		if key != firstKeyLabel {
+			log.Infof("mismatch Keys\n")
+			break
+		}
+		log.Infof("Zones Key: %s, Value: %s\n", key, value)
```
[the remark this answers is not in the document]
Comment (anathoodell) [reply controllers/csm_controller.go]: code here is revamped.

File: controllers/csm_controller.go [hunk 24]
```diff
@@ -1521,9 +1529,75 @@ func (r *ContainerStorageModuleReconciler) GetK8sClient() kubernetes.Interface {
...
+	for key, value := range zones {
+		if i == 0 {
+			firstKeyLabel = key
+			log.Infof("firstKeyLabel %s\n", firstKeyLabel)
+			i = i + 1
+			continue
+		}
+		if key != firstKeyLabel {
+			log.Infof("mismatch Keys\n")
+			break
```
[the remark this answers is not in the document]
Comment (anathoodell) [reply controllers/csm_controller.go]: yes code here is moved to the powerflex validation function (new PR).

File: controllers/csm_controller.go [hunk 25]
```diff
@@ -1521,9 +1529,75 @@ func (r *ContainerStorageModuleReconciler) GetK8sClient() kubernetes.Interface {
...
+	nodeList := &corev1.NodeList{}
+	log.Infof(" nodeList len %d", len(nodeList.Items))
+	var numberOfNodes = len(nodeList.Items)
+
+	for _, node := range nodeList.Items {
+		log.Infof(" nodeList name %s", node.Name)
+	}
+	zones, err := drivers.ExtractZones(ctx, cr, r.Client, namespace)
+	if err != nil {
+		log.Infof("ZoneValidation failed with error: %v", err)
```
[the remark this answers is not in the document]
Comment (anathoodell) [reply controllers/csm_controller.go]: done.

File: controllers/csm_controller.go [hunk 26]
```diff
@@ -1521,9 +1529,75 @@ func (r *ContainerStorageModuleReconciler) GetK8sClient() kubernetes.Interface {
...
+// Invalid zoning configurations:
+// 1. Some arrays have a zone label A, but one of the arrays has zone label B, or no label
+// 2. Array secret has zone label A, but one of the nodes has zone label B, or no label
+// 3. Some of the nodes have a zone label, but arrays have no zone labels.
+// 4. All of the arrays have a zone label, but nodes have no zone label.
+func (r *ContainerStorageModuleReconciler) ZoneValidation(ctx context.Context, cr *csmv1.ContainerStorageModule, namespace string) (bool, error) {
+	log := logger.GetLogger(ctx)
+	validZoneConfiguration := true
+	nodeList := &corev1.NodeList{}
+	log.Infof(" nodeList len %d", len(nodeList.Items))
```
[the remark this answers is not in the document]
Comment (anathoodell) [reply controllers/csm_controller.go]: removed.

**Pull request closed without merging**